### PR TITLE
Add reset gate

### PR DESCRIPTION
**Pull request by AthenaCaesura**
## Description

In the PR I introduce 2 new gates to `orquestra-quantum` the square root X gate (given by SX) and the RESET gate.

I have opted to implement RESET as a gate for 2 reasons:
 1. It simplifies the integration of reset into our other libraries as there are already robust testing frameworks for gates, but not for operations.
 2. Modifications like `Dagger` and `Controlled` should work with RESET even though it isn't unitary. They always produce matrices which can be applied to a circuit.

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [ ] I have updated documentation.

### CODE REVIEW

**Review by dexter2206 — COMMENTED** [review]
The SX implementation is totally fine.

However, I strongly recommend reconsidering implementation of Reset as operation. 

There are several reasons:
- If I understand the operation correctly it resets a single qubit to |0> state (a.k.a. performs deterministic measurement resulting in |0>). Thus, you cannot say that you can "always apply Reset matrix", unless by "apply" you mean something different than in the case of unitaries. Put differently, you apply gate by multiplying its matrix by the wavefunction, but the same cannot be said about the reset. Admittedly Reset has a matrix representation as a channel, but as explained above how it operates is quite dissimilar with unitary gates. We should not implement arbitrary stuff as a gate just because it has a matrix.
- Implementing Reset as gate destroys several abstractions. The biggest problem is with `Circuit.to_unitary()`. With other non-gate operations, this method will error-out so the user clearly knows that the circuit does not have a unitary representation. When you add a reset gate, this method will succeed and output a matrix U that is neither unitary nor describes the circuit action. Suppose someone is debugging their circuit, unknowingly gets such unitary and tries to figure out why multiplying it by the initial wavefunction results in a total nonsense.
- On the domain level, Reset is not a gate. Being consistent with domain language makes the code more readable and accessible to others, as well as easier to maintain.

The choice is yours. If you decide to stick with the gate approach, at least please remove `circuit.to_unitary()` method, before someone encounters a non-unitary unitary and starts pulling their hair debugging what went wrong.

**Comment by AthenaCaesura** [comment]
We need to functionality of putting `RESET` in a circuit. How could we do that if it's not an operation?

**Comment by AthenaCaesura** [comment]
I have re-implemented the Reset gate as an operation as an operation rather than a gate. I have also implemented it to throw an error if the user tries to compile into a single unitary. This gate is still useful because we can use it in benchq to convert between gates in `cirq` and `qiskit`.

**Review by max-radin — CHANGES_REQUESTED** [review]
Left a few comments about methods for `ResetOperation`. Otherwise looks good!

File: src/orquestra/quantum/circuits/_wavefunction_operations.py [hunk 1]
```diff
@@ -86,3 +86,44 @@ def free_symbols(self) -> Iterable[sympy.Symbol]:
...
+    def lifted_matrix(
+        self, amplitude_vector: ParameterizedVector
+    ) -> ParameterizedVector:
+        raise NotImplementedError(
+            "ResetOperation.apply is not implemented because the result would not be "
+            "a valid wavefunction."
+        )
+
+    def apply(self, amplitude_vector: ParameterizedVector) -> ParameterizedVector:
+        raise NotImplementedError(
```
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/orquestra/quantum/circuits/_wavefunction_operations.py:122]: No, it isn't feasable without using density matrix formalism (or quantum trajectories) 😞. The result in general wouldn't be a wavefunction, but a density matrix. Suppose you reset a qubit in a bell state: |00> + |11>. Then its partner will be in the maximally mixed state.

File: src/orquestra/quantum/circuits/_wavefunction_operations.py [hunk 2]
```diff
@@ -86,3 +86,44 @@ def free_symbols(self) -> Iterable[sympy.Symbol]:
...
+
+    def bind(self, symbols_map) -> "ResetOperation":
+        return self.replace_params(
+            tuple(sub_symbols(param, symbols_map) for param in self.params)
+        )
+
+    def replace_params(self, new_params: Tuple[Parameter, ...]) -> "ResetOperation":
+        return replace(self, params=new_params)
+
+    def lifted_matrix(
```
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/orquestra/quantum/circuits/_wavefunction_operations.py:113]: Done!

File: src/orquestra/quantum/circuits/_wavefunction_operations.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/orquestra/quantum/circuits/_wavefunction_operations.py:122]: Moreover following PEP specifications would be much more work than it's worth in this case. Specifically, PEP says you should delete the `apply`. But then it does not follow the `Operation` protocol which is required to put it in a circuit. So basically, following PEP here means rewriting how circuits are done from the ground up and frankly it isn't worth it just to add a dummy operation that doesn't do anything...

[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/orquestra/quantum/circuits/_wavefunction_operations.py:122]: As such, I'm just going to leave `apply` as throwing a `RuntimeError` for now and if we need this feature more we can rewrite how we do circuits then.

[the remark this answers is not in the document]
Comment (max-radin) [reply src/orquestra/quantum/circuits/_wavefunction_operations.py]: Makes sense, thanks!

**Review by max-radin — APPROVED** [review]

**Pull request merged**